### ISSUE TITLE
Counter based random number generators?

### ISSUE THREAD

**Issue by tdd11235813** [alpaka-group/alpaka#710]
I am looking for an efficient way to obtain the same pseudo random numbers being independent of the backend and parallelization. The focus is on parallel RNG, so it can be easily mapped to serial execution.

__Counter/Index based RNGs__ ([1](https://en.wikipedia.org/wiki/Counter-based_random_number_generator_%28CBRNG%29), [2](https://streamhpc.com/blog/2016-03-18/random-numbers-in-parallel-computing-generation-and-reproducibility-part-1/), [3](https://www.iaria.org/conferences2015/filesADVCOMP15/parallel-rng-advcomp15.pdf)) are also lightweight (stateless) and fast in the parallel environment, and therefore used in parallel Monte-Carlo etc.
Other RNGs also offer streams and sub-streams like MRG32k3a with "fast enough" jump aheads and also might offer better quality than CBRNGs, but have states, are slower (e.g., for CBRNG skip-ahead is O(1)) and might be more tricky to use it correctly in parallel environment.

With rocRAND/hipRAND and cuRAND we have RNGs suited for parallel reproducible random number generation, and they also include counter based RNGs. It looks like alpaka does not provide wrapper for it and the host side does not even have support for it.
[Random123](http://www.thesalmons.org/john/random123/) could be a library for this gap, because it offers CBRNGs and has CPU as well as CUDA and OpenCL backends, although CPU would be the interesting part for now.

Related to: #390 (unify RNG selection), #610 (Monte Carlo example), upcoming N-body project for mephisto (with real physics ;))

PS: Philox is an CBRNG, and here is [rocRAND's implementation](https://github.com/ROCmSoftwarePlatform/rocRAND/blob/master/library/include/rocrand_philox4x32_10.h)

Edit: just want to add that pregenerating random numbers and using them via a table-based approach is not useful, when you have to generate a huge number of them.

**Comment by j-stephan** [comment]
We discussed this in today's meeting. @sliwowitz is currently working on a separate RNG library on top of alpaka that will adress this issue. This is therefore WONTFIX and will be closed once the new RNG library is public.
